### PR TITLE
Merge neighbouring cells

### PR DESCRIPTION
**Pull request by betatim**
Merge two cells into one.

Couldn't find a way to hide the merge icon for the last cell 😢

L117 and L123 seem pretty hacky/not pretty to look at. Ideas/advice on good JS style?

### CODE REVIEW

**Comment by rgbkrk** [comment]
Oh great, I've been hoping for that. I'll check out the code thoroughly since you're asking.

**Comment by betatim** [comment]
Don't merge yet, there is a left over action that should have been deleted

File: src/notebook/actions/index.js [hunk 1]
```diff
@@ -164,6 +164,20 @@ export function createCellAppend(cellType) {
...
+    id,
+  };
+}
+
+export function mergeCellAfter(id) {
+  return {
+    type: constants.MERGE_CELL_AFTER,
+    id,
+  };
+}
```
Comment (rgbkrk) [inline src/notebook/actions/index.js:172]: 👍

File: src/notebook/components/cell/cell-creator-buttons.js [hunk 2]
```diff
@@ -36,7 +42,21 @@ export default class CellCreatorButtons extends React.Component {
...
+    }
+  }
+
   render() {
+    const mergeButton = (
+      <button onClick={this.mergeCell} title="merge cells">
+        <i className="material-icons" style={{ transform: 'rotate(90deg)' }}>
+          compare_arrows
+        </i>
+      </button>);
```
Comment (rgbkrk) [inline src/notebook/components/cell/cell-creator-buttons.js:54]: Ah cool, new button.

Comment (rgbkrk) [inline src/notebook/components/cell/cell-creator-buttons.js:54]: What happens when the cells are of different type?

Comment (betatim) [inline src/notebook/components/cell/cell-creator-buttons.js:54]: Right now they get merged, and it looks weird ;)

File: src/notebook/reducers/document.js [hunk 3]
```diff
@@ -103,6 +103,29 @@ export default {
...
+    const cellOrder = notebook.get('cellOrder');
+    const cellMap = notebook.get('cellMap');
+    const index = cellOrder.indexOf(id);
+    // do nothing if this the last cell
+    if (cellOrder.size === index + 1) {
+      return state;
+    }
+    const nextId = cellOrder.get(index + 1);
+    const source = `${cellMap.get(id).get('source')}
+${cellMap.get(nextId).get('source')}`;
```
Comment (rgbkrk) [inline src/notebook/reducers/document.js]: Would a `\n` style newline still work here or do you have to break it apart for backtick strings?

File: src/notebook/reducers/document.js [hunk 4]
```diff
@@ -103,6 +103,29 @@ export default {
...
   },
+  [constants.MERGE_CELL_AFTER]: function mergeCellAfter(state, action) {
+    const { id } = action;
+    const { notebook } = state;
+    const cellOrder = notebook.get('cellOrder');
+    const cellMap = notebook.get('cellMap');
+    const index = cellOrder.indexOf(id);
+    // do nothing if this the last cell
+    if (cellOrder.size === index + 1) {
+      return state;
```
Comment (rgbkrk) [inline src/notebook/reducers/document.js:114]: Good to cover this case. In addition, we can make sure the button doesn't appear in the UI for that.

**Comment by betatim** [comment]
Ok, the `mergeCellBefore` is gone.

File: src/notebook/reducers/document.js [hunk 5]
```diff
@@ -103,6 +103,29 @@ export default {
...
+    }
+    const nextId = cellOrder.get(index + 1);
+    const source = `${cellMap.get(id).get('source')}
+${cellMap.get(nextId).get('source')}`;
+
+    return {
+      ...state,
+      notebook:
+        commutable.removeCell(
+          commutable.updateSource(notebook, id, source),
```
Comment (rgbkrk) [inline src/notebook/reducers/document.js:124]: Looks like the right approach.

File: src/notebook/reducers/document.js (hunk 3 above)
Comment (betatim) [inline src/notebook/reducers/document.js]: Started with

```
const source = cellMap.get(id).get('source') + '\n' + cellMap.get(nextId).get('source');
```

but eslint doesn't like that and recommend this backtick style. Means you can't indent L118 though as that whitespace becomes part of the variable's contents.

**Comment by rgbkrk** [comment]
This reminds me that I need to get my split cell branch up. I still have it sitting around, haven't pushed since I got distracted with right-click context menus.

File: src/notebook/reducers/document.js (hunk 5 above)
Comment (betatim) [inline src/notebook/reducers/document.js:124]: It gets clunky when you also want to clear the output of the cell you are keeping. Actually not sure if we should clear it or not ... right now I went for keeping the output as the code for clearing it looked ugly.

File: src/notebook/reducers/document.js (hunk 3 above)
Comment (rgbkrk) [inline src/notebook/reducers/document.js]: Alternatively:

``` js
const source = cellMap.getIn([id, 'source']).concat('\n', cellMap.getIn([nextId, 'source']))
```

File: src/notebook/components/cell/cell-creator-buttons.js (hunk 2 above)
Comment (betatim) [inline src/notebook/components/cell/cell-creator-buttons.js:54]: We merge no matter what the type, just like jupyter does.

File: src/notebook/reducers/document.js (hunk 5 above)
Comment (rgbkrk) [inline src/notebook/reducers/document.js:124]: Ah, hmmm. Not sure of the right thing to do. Seems like you'd keep the output of the bottom cell?

Comment (betatim) [inline src/notebook/reducers/document.js:124]: jupyter keeps the output of the bottom cell. Not sure which one of the three options is best.

I'll ponder it while 😴

**Comment by betatim** [comment]
:shipit:'ing done. Merge at will.

**Comment by rgbkrk** [comment]
Merged the Merger!

**Pull request merged**